chapter/methodology.md

```diff
@@ -0,0 +1,1166 @@
+    It is a list of `LinearizationItem`s of unresolved type, appended as they are created during a depth-first traversal of the AST.
+  ~ During this phase, the `id` affected to a new item is always equal to its index in the array.
+  ~ The Building state also records the definitions in scope of each item in a separate mapping.
+
+post-processing phase:
+  ~ Once fully built, a Building instance is post-processed to get a `Completed` linearization.
+  ~ Although fundamentally still represented by an array, a completed linearization is optimized for search by positions (in the source file) thanks to sorting and the use of an auxiliary map from `id`s to the new index of items.
+  ~ Additionally, missing edges in the usage graph have been created and the types of items are fully resolved in a completed linearization.
+
+Type definitions of the `Linearization` as well as its type-states `Building` and `Completed` are listed in [@lst:nickel-definition-lineatization;@lst:nls-definition-building-type;@lst:nls-definition-completed-type].
+Note that only the former is defined as part of the Nickel libraries, the latter are specific implementations for NLS.
+
+
+```{.rust #lst:nickel-definition-lineatization caption="Definition of Linearization structure"}
+pub trait LinearizationState {}
+
+pub struct Linearization<S: LinearizationState> {
+    pub state: S,
+}
+```
+
+```{.rust #lst:nls-definition-building-type caption="Type Definition of Building state"}
+pub struct Building {
+    pub linearization: Vec<LinearizationItem<Unresolved>>,
+    pub scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+}
+impl LinearizationState for Building {}
+```
+
+```{.rust #lst:nls-definition-completed-type caption="Type Definition of Completed state" }
+pub struct Completed {
+    pub linearization: Vec<LinearizationItem<Resolved>>,
+    scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+    id_to_index: HashMap<ID, usize>,
+}
+impl LinearizationState for Completed {}
+```
+
+### Transfer from AST
+
+The NLS project aims to present a transferable architecture that can be adapted for future languages.
+Consequently, NLS faces the challenge of satisfying multiple goals
+
+
+<!-- what is more? -->
+
+To accommodate these goals NLS comprises three different parts as shown in [@fig:nls-nickel-structure].
+
+The `Linearizer` trait 
+  ~ acts as an interface between Nickel and the language server.
+    NLS implements a `Linearizer` specialized to Nickel which registers AST nodes and builds a final linearization.
+Nickel's type checking implementation
+  ~ was adapted to pass AST nodes to the `Linearizer`.
+    Modifications to Nickel are minimal, comprising only few additional function calls and a slightly extended argument list.
+A stub implementation
+  ~ of the `Linearizer` trait is used during normal operation.
+    Since most methods of this implementation are `no-op`s, the compiler should be able to optimize away all `Linearizer` calls in release builds.
```

Comment:
```suggestion
  ~ of the `Linearizer` trait is used during normal operation of the interpreter.
    Since most methods of this implementation are `no-op`s, the compiler is expected to be able to remove most `Linearizer` related method calls in optimized release builds.
    This promises minimal runtime impact incurred by the integration of lsp APIs.
```